libs/model/src/aggregates/community/LinkNamespace.command.ts

```diff
@@ -0,0 +1,169 @@
+import { type Command } from '@hicommonwealth/core';
+import * as schemas from '@hicommonwealth/schemas';
+import { BalanceSourceType } from '@hicommonwealth/shared';
+import { Transaction } from 'sequelize';
+import { z } from 'zod';
+import { models } from '../../database';
+import { mustExist } from '../../middleware/guards';
+import { emitEvent } from '../../utils';
+
+async function updateReferralCount(
+  referrer_address: string,
+  transaction: Transaction,
+) {
+  const referrer = await models.User.findOne({
+    include: [
+      {
+        model: models.Address,
+        where: { address: referrer_address },
+      },
+    ],
+    transaction,
+  });
+  referrer &&
+    (await referrer.update(
+      {
+        referral_count: models.sequelize.literal(`
+        (SELECT COUNT(DISTINCT referee_address) FROM "Referrals" 
+         WHERE referrer_address = ${models.sequelize.escape(referrer_address)})
+      `),
+      },
+      { transaction },
+    ));
+}
+
+async function setReferral(
+  namespace_address: string,
+  log_removed: boolean,
+  {
+    referrer_address,
+    referee_address,
+    timestamp,
+    eth_chain_id,
+    transaction_hash,
+  }: z.infer<typeof schemas.NamespaceReferral>,
+  transaction: Transaction,
+) {
+  const existingReferral = await models.Referral.findOne({
+    where: { namespace_address, referrer_address },
+    transaction,
+  });
+
+  // handle on-chain log removals
+  if (existingReferral) {
+    if (
+      existingReferral.transaction_hash === transaction_hash &&
+      existingReferral.eth_chain_id === eth_chain_id
+    ) {
+      if (log_removed) {
+        await existingReferral.destroy({ transaction });
+        await updateReferralCount(referrer_address, transaction);
+      }
+      return;
+    }
+    throw new Error(
+      `Found referral of namespace ${namespace_address} with mismatched chain details`,
+    );
+  }
+
+  await models.Referral.create(
+    {
+      namespace_address,
+      referrer_address,
+      referee_address,
+      eth_chain_id,
+      transaction_hash,
+      referrer_received_eth_amount: 0n,
+      created_on_chain_timestamp: BigInt(timestamp),
+    },
+    { transaction },
+  );
+  await updateReferralCount(referrer_address, transaction);
+}
+
+export function LinkNamespace(): Command<typeof schemas.LinkNamespace> {
+  return {
+    ...schemas.LinkNamespace,
+    auth: [],
+    body: async ({ payload }) => {
+      const { namespace_address, deployer_address, log_removed, referral } =
```

Comment:
We should probably have an `auth` option for "system actor only" so we don't accidentally expose these routes.

Reply:
We have a few commands already using systemActor in policy handlers... easy to find them and apply some extra layer of security... at the moment we don't expose them to the routers.